src/util/readaheadsamplebuffer.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+
+#include "util/indexrange.h"
+#include "util/samplebuffer.h"
+
+
+namespace mixxx {
+
+// A FIFO/LIFO sample buffer with fixed capacity and range checking.
+// It works best when consuming all buffered samples before writing
+// any new samples.
+//
+// This class is not thread-safe and is not intended to be used from
+// multiple threads!
+class ReadAheadSampleBuffer {
+  public:
+    ReadAheadSampleBuffer();
+    explicit ReadAheadSampleBuffer(
+            SINT capacity);
+    ReadAheadSampleBuffer(
+            const ReadAheadSampleBuffer& that)
+        : ReadAheadSampleBuffer(that, that.capacity()) {
+    }
+    ReadAheadSampleBuffer(
+            const ReadAheadSampleBuffer& that,
+            SINT capacity);
+    ReadAheadSampleBuffer(
+            ReadAheadSampleBuffer&&) = default;
+    virtual ~ReadAheadSampleBuffer() {}
+
+    ReadAheadSampleBuffer& operator=(
+            const ReadAheadSampleBuffer& that) {
+        *this = ReadAheadSampleBuffer(that); // copy ctor + move assignment
+        return *this;
+    }
+    ReadAheadSampleBuffer& operator=(
+            ReadAheadSampleBuffer&& that) {
+        swap(that);
+        return *this;
+    }
+    
+    void swap(
+            ReadAheadSampleBuffer& that);
+
+    // The maximum capacity of the buffer.
+    SINT capacity() const {
+        return m_sampleBuffer.size();
+    }
+
+    // Discards all buffered samples.
+    void clear();
+
+    bool empty() const {
+        return m_readableRange.empty();
+    }
+
+    // The number of samples that could be written instantly without
+    // internal reorganization, i.e. the remaining capacity of the
+    // buffer.
+    SINT writableLength() const {
+        return capacity() - m_readableRange.end();
```

Comment:
m_readableRange start always at 0?

Reply:
No. I added an additional comment that tries to clarify how the buffering internally works. Because it is not a full-featured ring buffer the full capacity only becomes available again after all written samples have been consumed. But that's sufficient for our use case.